man/man6/wlife.6
.TH WLIFE 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wlife \- Conway's game of life for W
.SH SYNOPSIS
wlife [-g \fIgeometry\fP]
.SH DESCRIPTION
.I Wlife
is a small program which serves two purposes: first it calculates and draws a
two-dimensional cellular automaton and second it shows you how to deal with
buttons and events.
.PP
You can set pixels by pressing the left mouse button and clear them with the
right one, if the simulation isn't actually running.
.PP
It now takes a geometry argument to work on variable fields. The positioning
part of geometry is not yet implemented.
.SH BUGS
xlife is faster...
.SH SEE ALSO
scan_geometry(3)
.SH LAST CHANGED
15th. Feb. 1996 by TeSche


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/wlife.c
/*
 * wlife.c, a part of the W Window System
 *
 * Copyright (C) 1994-1997 by Torsten Scherer
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- a Conway's Life 'game'
 *
 * CHANGES
 * ++TeSche, 12/95:
 * - added variable fields
 * ++TeSche, 01/96:
 * - new button concept
 * ++Eero, 10/97:
 * - use w_buttonqueryevent() instead of w_queryevent()
 */

#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <Wlib.h>


/*
 *
 */

typedef struct {
  int width, height;
  char *data;
} FIELD;


/*
 * some global variables
 */

static WWIN *glob_win, *buttonQuit, *buttonCont, *buttonInt;
static int glob_yoffset, glob_terminate;


/*
 * allocField(): allocate a field and initialize its cells to zero. internally
 * we make the field two cells bigger in each dimension and 'mirror' over the
 * border when a cell at the edge of the real field is set. this way we can
 * save a lot of boundary checks when calculating the neighbours of a pixel.
 */

static FIELD *allocField(int width, int height)
{
  FIELD *ret;
  int size;

  if (!(ret = malloc(sizeof(FIELD))))
    return NULL;

  width += 2;
  height += 2;

  size = width * height;
  if (!(ret->data = malloc(size))) {
    free(ret);
    return NULL;
  }

  memset(ret->data, 0, size);

  ret->width = width;
  ret->height = height;

  return ret;
}


static inline void copyField(FIELD *dest, FIELD *source)
{
  memcpy(dest->data, source->data, source->width * source->height);
}


/*
 * indices to these functions here are within the original range!
 */

#define XLO 0x0001
#define XHI 0x0002
#define YLO 0x0004
#define YHI 0x0008

static inline void set(register FIELD *dest,
		register int x,
		register int y,
		register char val)
{
  register int idx = (y + 1) * dest->width + x + 1;
  register char *data = dest->data;
  register int edges;

  data[idx] = val;

  if ((edges = (!x ? XLO : 0) | ((x == (dest->width - 3)) ? XHI : 0) |
       (!y ? YLO : 0) | ((y == (dest->height - 3)) ? YHI : 0))) {

    switch (edges) {

      case XLO | YLO:
        data[idx + dest->width - 2] = val;
	idx += (dest->height - 2) * dest->width;
	data[idx] = val;
        data[idx + dest->width - 2] = val;
	break;

      case YLO:
	idx += (dest->height - 2) * dest->width;
	data[idx] = val;
	break;

      case XHI | YLO:
	data[dest->width] = val;
	idx += (dest->height - 2) * dest->width;
	data[idx] = val;
	idx -= dest->width - 2;
	data[idx] = val;
	break;

      case XHI:
	idx -= dest->width - 2;
	data[idx] = val;
	break;

      case XHI | YHI:
	data[0] = val;
	data[dest->width - 2] = val;
	idx -= dest->width - 2;
	data[idx] = val;
	break;

      case YHI:
	data[x + 1] = val;
	break;

      case XLO | YHI:
	data[1] = val;
	data[dest->width - 1] = val;
	idx += dest->width - 2;
	data[idx] = val;
	break;

      case XLO:
	idx += dest->width - 2;
	data[idx] = val;
	break;
    }
  }
}


static inline char isSet(FIELD *dest, int x, int y)
{
  return dest->data[(y + 1) * dest->width + x + 1];
}



static inline int numNeighbours(register FIELD *dest,
			 register int x,
			 register int y)
{
  register int bpl = dest->width;
  register char *data = dest->data + y * bpl + x;   /* upper left */
  register int n = 0;

  n = *data++;
  n += *data++;
  n += *data;
  data += bpl;
  n += *data;
  data += bpl;
  n += *data--;
  n += *data--;
  n += *data;
  data -= bpl;
  n += *data;

  return n;
}


/*
 * the main 'life' routine
 */

static void do_life(FIELD *field, int width, int height)
{
  WEVENT *wevent;
  int end, x, y;
  static FIELD *copy = NULL;

  if (!copy) {
    if (!(copy = allocField(width, height))) {
      fprintf(stderr, "error: out of memory\n");
      exit(-1);
    }
  }

  w_setmode(glob_win, M_INVERS);

  end = 0;
  while (!end) {

    /* compute new field */
    for (y=0; y<height; y++) {
      for (x=0; x<width; x++) {
	switch (numNeighbours(field, x, y)) {
	  case 2:
	    set(copy, x, y, isSet(field, x, y));
	    break;
	  case 3:
	    set(copy, x, y, 1);
	    break;
	  default:
	    set(copy, x, y, 0);
	    break;
	  }
      }
    }

    /* update screen */
    for (y=0; y<height; y++) {
      for (x=0; x<width; x++) {
	if (isSet(copy, x, y) != isSet(field, x, y)) {
	  w_pbox(glob_win, 4*x, 4*y + glob_yoffset, 4, 4);
	}
      }
    }

    /* update field */
    copyField(field, copy);

    /* query interrupts */
    if ((wevent = w_querybuttonevent(NULL, NULL, NULL, 0))) {
      if (wevent->type == EVENT_BUTTON) {
	if (wevent->win == buttonInt) {
	  end = 1;
	} else if (wevent->win == buttonQuit) {
	  end = 1;
	  glob_terminate = 1;
	}
      }

      if ((wevent->type == EVENT_GADGET) &&
	  ((wevent->key == GADGET_EXIT) || (wevent->key == GADGET_CLOSE))) {
	end = 1;
	glob_terminate = 1;
      }
    }
  }
}


/*
 * usage()
 */

static void usage(void)
{
  fprintf(stderr, "usage: wlife [-g <geometry>]\n");
  exit(-1);
}


/*
 * guess what...
 */

int main(int argc, char *argv[])
{
  short width = 64, height = 40, winw, winh, cx, cy;
  WEVENT *wevent;
  WFONT *wfont;
  FIELD *field;

  signal(SIGTTOU, SIG_IGN);

  /*
   *
   */

  switch (argc) {
    case 3:
      width = 64;
      height = 40;
      scan_geometry(argv[2], &width, &height, NULL, NULL);
      if (width < 64) {
	width = 64;
      }
      if (height < 40) {
	height = 40;
      }
      /* fall thru */
    case 1:
      break;

    default:
      usage();
  }

  /*
   *
   */

  if (!w_init()) {
    fprintf(stderr, "fatal: wlife can't connect to wserver\n");
    exit(-1);
  }

  if (!(wfont = w_loadfont("fixed", 8, 0))) {
    fprintf(stderr, "error: wlife can't load font\n");
    exit(-1);
  }

  glob_yoffset = wfont->height + 5;
  winw = width * 4;
  winh = height * 4 + glob_yoffset;

  if (!(field = allocField(width, height))) {
    fprintf(stderr, "error: out of memory\n");
    exit(-1);
  }

  if (!(glob_win = w_create(winw, winh, W_MOVE | W_TITLE | W_CLOSE | EV_MOUSE))) {
    fprintf(stderr, "error: wlife can't create window\n");
    exit(-1);
  }

  w_settitle(glob_win, " wlife, (C) 12/95 by TeSche ");

  if (w_open(glob_win, UNDEF, UNDEF) < 0) {
    w_delete(glob_win);
    fprintf(stderr, "error: wlife can't open window\n");
    exit(-1);
  }

  buttonQuit = w_createButton(glob_win,   0, 0, 50, 9);
  buttonCont = w_createButton(glob_win,  58, 0, 90, 9);
  buttonInt =  w_createButton(glob_win, 156, 0, 90, 9);
  if (!buttonQuit || !buttonCont || !buttonInt) {
    fprintf(stderr, "error: wlife can't create buttons\n");
    exit(-1);
  }

  w_centerPrints(buttonQuit, wfont, "quit");
  w_centerPrints(buttonCont, wfont, "continue");
  w_centerPrints(buttonInt, wfont, "interrupt");

  w_showButton(buttonQuit);
  w_showButton(buttonCont);
  w_showButton(buttonInt);

  glob_terminate = 0;
  while (!glob_terminate) {

    if ((wevent = w_querybuttonevent(NULL, NULL, NULL, -1))) switch (wevent->type) {

      case EVENT_BUTTON:
        if (wevent->win == buttonCont) {
	  do_life(field, width, height);
	} else if (wevent->win == buttonQuit) {
	  glob_terminate = 1;
	}

      case EVENT_MPRESS:
	if (wevent->y > glob_yoffset) {
	  cx = wevent->x / 4;
	  cy = (wevent->y - glob_yoffset) / 4;
	  switch (wevent->key) {

	    case BUTTON_LEFT:
	      set(field, cx, cy, 1);
	      w_setmode(glob_win, M_DRAW);
	      w_pbox(glob_win, 4*cx, 4*cy + glob_yoffset, 4, 4);
	      break;

	    case BUTTON_RIGHT:
	      set(field, cx, cy, 0);
	      w_setmode(glob_win, M_CLEAR);
	      w_pbox(glob_win, 4*cx, 4*cy + glob_yoffset, 4, 4);
	      break;
	  }
	}
	break;

      case EVENT_GADGET:
	if ((wevent->key == GADGET_EXIT) ||
	    (wevent->key == GADGET_CLOSE)) {
	  glob_terminate = 1;
	}
	break;
      }
  }

  return 0;
}
